File: erlang/kmeans.erl
-module(kmeans).
-export([run/3]).
-compile(inline).

run(Xs, N, Iters) ->
    InitCentroids = lists:sublist(Xs, N),
    Step = fun(_, Centroids) ->
                   [average(X) || X <- clusters(Xs, Centroids)]
           end,
    FinalCentroids = lists:foldl(Step, InitCentroids, lists:seq(1, Iters)),
    clusters(Xs, FinalCentroids).

divide({Px,Py}, K) ->
    {Px/K, Py/K}.

add({Px1, Py1}, {Px2, Py2}) ->
    {(Px1+Px2), (Py1+Py2)}.

sub({Px1, Py1}, {Px2, Py2}) ->
    {(Px1-Px2), (Py1-Py2)}.

sq(X) ->
    X*X.

modulus({Px, Py}) ->
    math:sqrt((sq(Px) + sq(Py))).

dist(P1, P2) ->
    modulus(sub(P1,P2)).

average(Q) ->
    divide(sum(Q),length(Q)).

sum(L) ->
    Add = fun(X, Acc) -> add(X, Acc) end,
    lists:foldl(Add, {0.0, 0.0}, L).

closest(P, Centroids) ->
    element(2, lists:min([{dist(P, C), C} || C <- Centroids])).

clusters(Xs, Centroids) ->
    groupBy(Xs, fun(X) -> closest(X, Centroids) end).

groupBy(L, Fn) ->
    Group = fun(X, Dict) ->
                    Add = fun(T) -> [X|T] end,
                    dict:update(Fn(X), Add, [X], Dict)
            end,
    Dict = lists:foldl(Group, dict:new(), L),
    [ V || {_,V} <- dict:to_list(Dict)].

File: fsharp/kmeans.fs
module Kmeans

type Point = { x: double; y: double }

let sq (x: double) = x*x

type Point with
  static member (+) (p1: Point, p2: Point) =
    { x = p1.x + p2.x; y = p1.y + p2.y }

  static member (-) (p1: Point, p2: Point) =
    { x = p1.x - p2.x; y = p1.y - p2.y }

  static member (*) (p1: Point, p2: Point) =
    { x = p1.x * p2.x; y = p1.y * p2.y }

  static member (/) (p: Point, d: double) =
    { x = p.x / d; y = p.y / d }

  static member Zero = { x = 0.0; y = 0.0 }

  member this.modulus = sqrt ( sq(this.x) + sq(this.y))

let dist (p1: Point, p2: Point) = (p1 - p2).modulus

let average (xs: Point list) = (xs |> List.sum) / (xs.Length |> double)

let closest (p: Point, xs: Point list) =
  xs |> List.minBy( fun x -> dist (p, x) )

let clusters (xs: Point list, centroids: Point list) =
  xs |> List.groupBy( fun x -> closest ( x, centroids)) |> List.map snd

let rec run (xs: Point list, centroids: Point list, iters: int) =
  match iters with 
    | 1 -> 
      let finalCentroids = clusters ( xs, centroids) |> List.map average
      //printfn "Final centroids are %A" finalCentroids
      finalCentroids
    | _ -> run (xs, clusters ( xs, centroids) |> List.map average, iters - 1)

let start (xs: Point list, iters: int, n: int) =
  run (xs, xs |> List.take n, iters)
